### ISSUE TITLE
Merge permissions

### ISSUE THREAD

**Issue by sjakobi** [haskell-nix/hnix#598]
Currently it seems that, to merge a PR, it must first be approved by a different collaborator. [Apparently these approvals are also invalidated when the PR branch is rebased](https://github.com/haskell-nix/hnix/pull/574#issuecomment-634817180).

IMHO, for most PRs, this causes more fuss than appropriate. I fully trust @Anton-Latukha and @sorki to be able to judge whether a patch could use a second pair of eyeballs or whether it is simple enough to be merged without much ado.

So I'm wondering whether the current merge policy could be relaxed a bit.

If we want some formal guidelines regarding who may merge under which circumstances, I think we should maybe rather write them down in a `CONTRIBUTING.md`, as [for example the Dhall  project](https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#how-do-changes-get-approved) has done with great success.

Thoughts?

**Comment by Anton-Latukha** [comment]
Yes.

The process is too stingy currently

I am still thinking about this.

I prefer a structured process that resolves things properly itself, and so it does not need `in the head` rules.
Since `in the head` law and rules are imagination anyway, that is only perceived as true due to its attachment to social structure. Those `keep in the mind` restricting causes psychological suffering, internal conflict and so people want to break them. We can always create new laws to solve old laws, and in the end, we would need lawyers to buy and eat dinner.

I have the too small experience to the project maintenance to put things simply.

The cross-review seems inevitable in the big multilevel community. Between 3 people, like now - it looks a lot like a tight population bottleneck and it is an era when nepotism between us is inevitable and even more - needs to be fully embraced.

**Comment by Anton-Latukha** [comment]
Proposition:

Maybe the nearest optimal answer is - allow maintainers themself to review their own PRs currently, or lift the review requirement from maintainers, so they would themself cross-call for review themselves.

Arguments:

---

1. The requirement to sync to the `master` branch and pass the CI - are the main requirements. They should be in place and can not be lifted.

<details><summary>wrapper</summary>
<p>


Code should always be consistent, it is the main most important property of the code and the good code. Code should be consistent so it works and be reliable. - That is why syncing to `master` and passing CI builds & tests are the main requirement and that should be so.

</details>


---

2. The review - is a bonus process that allows us to discuss the changes. For Haskell projects the review may increase the quality easy PRs a bit, and review is needed for major changes. The review process in itself is a manual, ineffective in the time*effort consuming process that produces lesser results comparing to actually doing the work and relying on Haskell type system & CI. So we need to use it only for changes that need review.

<details><summary>wrapper</summary>
<p>

Since there is a lot of:
  * easy fruit work currently in the program,
  * and a lot of cosmetics and clean-up,
  * and I plan to document the code structure through #555 
 
Those are questions that strange to pass through the code review. Since:
  * there is nothing to review in them,
  * or the changes are too simple coming from the person that you know is qualified for much more, it is embarrassing to be required to review trivial changes
  * code has no documentation, it needs Haddock coverage and proper documentation. The absence of documentation is always (almost strictly) worse than having somewhat sane documentation.


</details>

---

3. There is a lot to document. And requiring review process on the documentation - creates a big unending recursive loop of work that requires a lot of effort, and never terminates.


<details><summary>wrapper</summary>
<p>

  * Documentation is not a code, PRs of documentation has no code, so there is nothing that changes the internal consistency/features of the code. There is no code - so there is nothing to review, except nitpick the typos/phrasing in the text.
  * Having docs with typos is better than never merging and having the documentation. So it is better to merge the docs and then fix them, over to block the docs in the review loop.
  * Documentation PRs with review requirement tent to go into the excessive recursive loop never terminates:
    * be embarrassed to bother people with the requested review, so you wait maybe someone would review themselves
    * time window of waiting
    * asked to review
    * awaiting review
    * your docs were consistent with the `master`, but now something merged into `master` - so now your documentation placing is inconsistent with the `master`
    * or GitHub simply invalidated the ability merge the documentation, because you need to sync rebase to `master`, and then pass the CI.
    * maintainer came, and as required effort into reading your documentation - reviewed it
    * rephrasing process
    * rebase to `master`
    * waiting on CI (~0.5-1-2 hour)
    * CI build tests (yes, for documentation, Nix sees everything, you changed the `sha` - so, lets referential transparently build it and cache the result that would never be used from cache anymore)
    * Test passed, since you made changes or made rebase - GitHub invalidates review, and it now requires new review to be able to merge
    * you arrived at step 1. GOTO Step 1 (be embarrassed to ask and bother with a review of documentation)...
    * ... never terminates, since there is always a possibility for rephrasing and introducing new errors/typos during rephrasing.
  * Some nearest examples of recursive loops that documentation and requirement of review creates: #556 #583. There is no questions to the code, in one - there is no code at all, but the recursive loop is there. I need to make a rebase/test/review loop because of a review on docs that introduces desyndc with `master` brunch. I and others could solve those question very quickly, if we were able to easily merge the doc changes into `master` by themselves.
  * We as programmers and people understand that loop above is created for the code (CI builds are a hint). And text peer review needed for articles, journals and books. And internal code documentation is a bonus to the code, not so much a feature.
  * Currently HNix has broken (absent) documentation on Hackage, and seems people are pretty silent about that, nobody ran to check it or control it, so why in PRs we minutia the documentation process.

Back in a day, I wrote an article in Serokell that would be going one of the first articles to the company blog. It was an easy theme and article also. Review and recommendations on how to make a basic setup for the Haskell project in Travis (using Stack/Cabal at the time, because most of Haskellers were using only them at the time).
The theme is very simple. But we were fresh into the process, and very picky about everything. The recursive review process went for months, it was so long that I got out of the company and so we never published that article. Dima was the technical reviewer. Those ideas and work to organize the setups helped to push Kowainik into creating the `summoner`, Dima was always great in the structuring things through tooling, he created a lot of useful tooling.

  * So we see code review and the code documentation for us should have different approaches
  * Documentation philosophy of the creators of the Wikis: https://en.wikipedia.org/wiki/Wiki#Trust_and_security
  Prioritize the easiness for the community to make changes over. So the community would solve their mistakes themselves and faster evolution of documentation results into better documentation.
  * The **requirement of the review directly creates a major paradigm clash with the main goal and philosophy of open documentation**: https://archive.org/details/isbn_9780201714999/page/16/mode/2up


</details>

---

4. There is a lot of low hanging fruit to do


<details><summary>wrapper</summary>
<p>

Dependency clean-up, code clean-up, module imports clean-up, increase module import explicitness (there are currently a function collisions in the code and its import, due to implicitness), cosmetic changes, minor refactoring...


</details>

---

So, since code needs to pass through the CI successfully, and we can make CI make a thorough check. And Haskell already a very big quality check-in in itself.
We can disable the requirement of the review process, or allow to review own PRs and allow maintainers to merge code themselves relying on trusting them.
We would raise the CI strictness gradually.

And three current maintainers know when to ask for dialog or ask to review the changes, and they care about their reputation in our project and community - so they themselves would ask for reviews on some moderate changes, and free one another from useless petty minutia by being lazy and on CI passing - merge the cosmetics and easy changes themself.

---

Results of the statement:

1. We have a pile of easy work and a small number of people, so we would allow them to work.

2. We would allow for fast progress and all useful easy work that maintainers (that are also most active contributors currently) do - to be quickly merged. The project needs a lot of clean-up, organization, and documentation structuring and coverage of the code, and covering of a lot of `warn` messages and rapid development overall to sync up with Nix.

3. People know when to ask others for review.
And contributors anyway always need to pass the maintainer review to merge.

4. We always can reintroduce the requirements in one click.

But I make a prognosis it would be long until we would need constant cross-review.

When the community would be large, and the HNix would be working and important project that many people rely on - so it would need to be in the efficient optimized state always, even in `master`, and we would claim production readiness - then we may need required cross-review.

Currently required cross-maintainer-review is only a constraint that exponentially multiplies the constant inconvenient loops and effort of going through the loops to merge the very simple and easy contribution (cosmetics and documentation PRs are best examples of the ridiculousness syndrome).

We required to wait a lot, rebase cosmetics a lot and to hold a discussion on everything, and to stagger in the rebase/CI/review/loops for petty things. When what we need is to have discussions and reviews only on what is important and trust one another to be effective.

Trust is effectiveness, bureaucracy is a torpor. The unrequired bureaucracy equivalent with the security theater.

So if we can be free from this bureaucracy constraint, we need to remove the constraint. That would align the process to the real-world, it current circumstances and cases. So the small maintainer/contributor community can more effectively work on the project.

**Comment by sjakobi** [comment]
Yeah, I've been a bit frustrated with rubber-stamp-approving your Nix- and CI-changes which I generally didn't have a good understanding of, or even none at all. In order to allow you to work uninhibitedly, I'd like to ask @jwiegley whether he could give you permissions to merge without approval from others.

Regarding changes to the core library, I'd currently prefer to be kept in the loop a bit. There hasn't been a lot of work on that yet though.

@sorki What do you think about all this.

**Comment by sorki** [comment]
I've checked the Nix/CI changes briefly and they were looking good, you've just beat me to approving them :+1: 

I agree that the requirement for review and re-reviews from maintainers is unnecessarily strict especially in the presence of CI.

**Comment by Anton-Latukha** [comment]
Indeed as you said the "review and re-reviews".

I'd opted-in into "easy reviews", if there was no need for calling/coming to review ahead of time, calling several reviewers at the same time so someone would come, waiting for review, on syncs to master or minor changes in PRs - re-reviews, re-re-reviews, and be notified and have a responsibility to come, and come for a lot of those false alarms and placebo reviews.

I understand the **_sjakobi_** in https://github.com/haskell-nix/hnix/issues/598#issuecomment-637904326, mostly solely he responded to my PRs in last days, and I am grateful.

I think reviews for maintainers should be an optional feature. As I said, the CI is and should and going to be the major check already.
The mandatory cross-reviews are debatable when the project inside the community is in a very good shape works good and used widely, currently there no of such conditions, the contrary conditions are true.

**Comment by jwiegley** [comment]
I've removed branch protections for now. Any maintainer or contributor can choose to merge. If we want to restore protections, please do so through the Settings, or let me know what you'd like the end user behavior to be.

**Comment by Anton-Latukha** [comment]
Thank you.

Main technical checks and tests are always up, all PRs need to pass guarding CI, it has Source Distribution and strict build checks now. Now we can more casually develop.

**Comment by Anton-Latukha** [comment]
The report topic is resolved.
